### ISSUE TITLE
how to use bindgen-generated types in a cxx rust callback?

### ISSUE THREAD

**Issue by ian-p-cooke** [dtolnay/cxx#894]
Hi,

I'm trying to use [PcapPlusPlus](https://pcapplusplus.github.io/) from Rust and am having trouble with every combination of bindgen, cxx, and auto-cxx... I think I'm pretty close with my latest bindgen+cxx approach but I'm stuck with an error that I can't figure out how to get past.  
```Rust
  error[cxxbridge]: unsupported type
     ┌─ src/main.rs:35:54
     │
  35 │         fn onMessageReadyCallback(side: i8, tcpData: &crate::ffi_pcpp::TcpStreamData, userCookie: *mut ::std::os::raw::c_void);
     │                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ unsupported type
```
I've boiled the problem down to this demo: https://github.com/ian-p-cooke/cxx-issue

Is it possible for a C++ library to invoke a function defined in Rust via a function pointer and pass Rust a reference to a C++ class that you created bindgen binding for?

what I need is a C++ function with the signature 
```C++
typedef void(*       OnTcpMessageReady) (int8_t side, const TcpStreamData &tcpData, void *userCookie)
```
which I'm attemting to do by using cxx to bind to a Rust function with signature
```Rust
fn onMessageReadyCallback(side: i8, tcpData: &crate::root::pcpp::TcpStreamData, userCookie: *mut ::std::os::raw::c_void) {
````
where `crate::root::pcpp::TcpStreamData` was generated by bindgen.

P.S.

I know the c_void mapping doesn't work either so that's my next problem.

**Comment by dtolnay** [comment]
It should work if you do:

```diff
diff --git a/include/pcap_reader.hpp b/include/pcap_reader.hpp
index 3fc1267..6c55caa 100644
--- a/include/pcap_reader.hpp
+++ b/include/pcap_reader.hpp
@@ -2,6 +2,7 @@
 #include "rust/cxx.h"
 
 struct UserCookie;
+using c_void = void;
 
 void read_pcap(rust::Str fileName,

diff --git a/src/main.rs b/src/main.rs
index e66a894..d28f565 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -14,17 +14,9 @@ unsafe impl ExternType for root::pcpp::ConnectionData {
     type Kind = cxx::kind::Opaque;
 }
 
-#[cxx::bridge(namespace = "pcpp")]
-mod ffi_ppcp {
-    extern "C++" {
-        type TcpStreamData = crate::root::pcpp::TcpStreamData;
-        type ConnectionData = crate::root::pcpp::ConnectionData;
-    }
-}
-
-struct UserCookie;
+pub struct UserCookie;
 
-fn onMessageReadyCallback(side: i8, tcpData: &crate::root::pcpp::TcpStreamData, userCookie: *mut ::std::os::raw::c_void) {
+fn onMessageReadyCallback(side: i8, tcpData: &crate::root::pcpp::TcpStreamData, userCookie: *mut ffi::c_void) {
     println!("onMessageReadyCallback!");
 }
 
@@ -32,11 +24,19 @@ fn onMessageReadyCallback(side: i8, tcpData: &crate::root::pcpp::TcpStreamData,
 mod ffi {
     extern "Rust" {
         type UserCookie;
-        fn onMessageReadyCallback(side: i8, tcpData: &crate::ffi_pcpp::TcpStreamData, userCookie: *mut ::std::os::raw::c_void);
+        unsafe fn onMessageReadyCallback(side: i8, tcpData: &TcpStreamData, userCookie: *mut c_void);
+    }
+
+    #[namespace = "pcpp"]
+    extern "C++" {
+        include!("demo/include/pcapplusplus/TcpReassembly.h");
+        type TcpStreamData = crate::root::pcpp::TcpStreamData;
+        type ConnectionData = crate::root::pcpp::ConnectionData;
     }
 
     extern "C++" {
         include!("demo/include/pcap_reader.hpp");
+        type c_void;
         unsafe fn read_pcap(fileName: &str,
                             userCookie: &mut UserCookie,
```

**Comment by ian-p-cooke** [comment]
thank you!  I did keep banging on it until I came up with https://github.com/ian-p-cooke/cxx-issue/commit/8629228641070d0da695b299d6fd7d2d0f05c358 
I like your solution to some things better though so in my main project I'll incorporate yours.
I found the `namespace` attribute in some issue which was the 'aha!' moment for me.
I never read https://cxx.rs/attributes.html *sigh* that's what I get for skimming docs.

thank you also for showing me how to handle the void problem.
